Enhance the find method

Tim-Siu opened the pull request:
- Implemented find functionalities for both persons and companies
- Added support for searching with tags

Note that I discover bugs in command parser for other commands. ArrayOutOfBounds bug is not caught when performing trimming to get the person/company argWord.

Tim-Siu committed:
Enhance the find method (1/n)

- Implemented FindCommand class
- Implemented enhanced predicate

After the enhancement is finished, the SOCareers will support finding persons or companies with names and (new!) tags. The syntax is updated to reflect the team's discussion as well.

Tim-Siu committed:
Enhance the find method (2/n)

- Implemented FindPersonCommand class and FindCompanyCommand class
- Implemented relevant test classes

After the enhancement is finished, the SOCareers will support finding persons or companies with names and (new!) tags. The syntax is updated to reflect the team's discussion as well.

oeggy03 commented on src/main/java/seedu/address/logic/commands/findcommands/FindPersonCommand.java:
```diff
@@ -0,0 +1,49 @@
+     * Creates a FindPersonCommand to find persons with the specified {@code NameAndTagContainKeywordsPredicate}
+     */
+    public FindPersonCommand(NameAndTagContainKeywordsPredicate predicate) {
+        assert predicate != null;
```
This is okay but why not use requireNonNull?

Tim-Siu replied:
I will change it to requireNonNull but honestly I don't know what is the difference haha

oeggy03 replied:
Same, I think it does the same thing haha

better abstraction maybe

oeggy03 commented on src/main/java/seedu/address/logic/commands/findcommands/FindCompanyCommand.java:
```diff
@@ -0,0 +1,51 @@
+     * with the specified {@code CompanyNameAndTagContainKeywordsPredicate}
+     */
+    public FindCompanyCommand(CompanyNameAndTagContainKeywordsPredicate predicate) {
+        assert predicate != null;
```
This is okay but why not use requireNonNull?

oeggy03 commented on src/main/java/seedu/address/model/company/CompanyNameAndTagContainKeywordsPredicate.java:
```diff
@@ -0,0 +1,68 @@
+    public boolean test(Company company) {
+        return companyNameKeywords.stream()
+            .anyMatch(keyword -> StringUtil.containsWordIgnoreCase(company.getCompanyName().fullName, keyword))
+            // Previously, the method was checking for keyword matches using company.getTags().toString(),
```
Thanks for the comments :) it makes the code much clearer and easier to read

oeggy03 commented on src/main/java/seedu/address/model/company/CompanyNameAndTagContainKeywordsPredicate.java:
```diff
@@ -0,0 +1,68 @@
+            // The updated approach streams through each individual tag of a Company and checks
+            // if any of them match the keywords. This ensures that we accurately check
+            // each tag against each keyword, providing a more precise match.
+            || company.getTags().stream()
```
Yep I think this will work

oeggy03 commented on src/test/java/seedu/address/model/person/NameAndTagContainKeywordsPredicateTest.java:
```diff
@@ -0,0 +1,105 @@
+            + "{nameKeywords=" + nameKeywords + ", tagKeywords=" + tagKeywords + "}";
+        assertEquals(expected, predicate.toString());
+    }
+}
```
I think that the tests look comprehensive

oeggy03 reviewed (commented):
Hi I think FindCommandParser is missing!

Let me know if I missed it out somewhere

Tim-Siu committed:
Enhance the find method (3/3)

- Finished implementing find feature

Note that I discover bugs in command parser for other commands. ArrayOutOfBounds bug is not caught when performing trimming to get the person/company argWord.

Tim-Siu commented:
> Hi I think FindCommandParser is missing!
> 
> Let me know if I missed it out somewhere

Indeed not there yet. This is in my latest commit🤩

Tim-Siu committed:
Enhance the find method (4/4)

- Made relevant changes to the problems pointed out

Note that I discover bugs in command parser for other commands. ArrayOutOfBounds bug is not caught when performing trimming to get the person/company argWord.

oeggy03 reviewed (approved):
Really good job! Another step forward for SOCareers :D
